**Sample ID**: 124

**Query**:

I want to return everything I just bought except for the coffee machine.

**DB Type**: Base Case

**Case Description**:

The user, Fatima Wilson (email: fatima.wilson5721@example.com), is requesting to return specific items from her recent order, #W5272531. She wants to keep the coffee machine but return all other items from that purchase. The items to be returned are identified by the IDs: '7228247242', '2698416822', '8098621301', and '3320557165'. The refund should be credited to her payment method, credit_card_6824399.





```
<multiturn info>
User Name: Fatima Wilson (Information Gathering)
User Email: fatima.wilson5721@example.com (Information Gathering)
Item to Keep: The coffee machine (Information Gathering)
</multiturn info>
```

**Global/Context Variables:**


**APIs:**

- retail


# Set Up

## Download relevant files

In [ ]:
import io
import os
import sys
import zipfile
import shutil
import re
from google.colab import auth
from googleapiclient.discovery import build
from googleapiclient.http import MediaIoBaseDownload

# Version to download
VERSION = "0.1.1"  # Version of the API

# Define paths
CONTENT_DIR = '/content'
APIS_DIR = os.path.join(CONTENT_DIR, 'APIs')
DBS_DIR = os.path.join(CONTENT_DIR, 'DBs')
SCRIPTS_DIR = os.path.join(CONTENT_DIR, 'Scripts')
FC_DIR = os.path.join(CONTENT_DIR, 'Schemas')
ZIP_PATH = os.path.join(CONTENT_DIR, f'APIs_V{VERSION}.zip')

# Google Drive Folder ID where versioned APIs zip files are stored
APIS_FOLDER_ID = '1QpkAZxXhVFzIbm8qPGPRP1YqXEvJ4uD4'

# List of items to extract from the zip file
ITEMS_TO_EXTRACT = ['APIs/', 'DBs/', 'Scripts/', 'Schemas/']

# Clean up existing directories and files
for path in [APIS_DIR, DBS_DIR, SCRIPTS_DIR, FC_DIR, ZIP_PATH]:
    if os.path.exists(path):
        if os.path.isdir(path):
            shutil.rmtree(path)
        else:
            os.remove(path)

# Authenticate and create the drive service
auth.authenticate_user()
drive_service = build('drive', 'v3')

# Helper function to download a file from Google Drive
def download_drive_file(service, file_id, output_path, file_name=None, show_progress=True):
    """Downloads a file from Google Drive"""
    destination = output_path
    request = service.files().get_media(fileId=file_id)
    with io.FileIO(destination, 'wb') as fh:
        downloader = MediaIoBaseDownload(fh, request)
        done = False
        while not done:
            status, done = downloader.next_chunk()
            if show_progress:
                print(f"Download progress: {int(status.progress() * 100)}%")


# 1. List files in the specified APIs folder
print(f"Searching for APIs zip file with version {VERSION} in folder: {APIS_FOLDER_ID}...")
apis_file_id = None

try:
    query = f"'{APIS_FOLDER_ID}' in parents and trashed=false"
    results = drive_service.files().list(q=query, fields="files(id, name)").execute()
    files = results.get('files', [])
    for file in files:
        file_name = file.get('name', '')
        if file_name.lower() == f'apis_v{VERSION.lower()}.zip':
            apis_file_id = file.get('id')
            print(f"Found matching file: {file_name} (ID: {apis_file_id})")
            break

except Exception as e:
    print(f"An error occurred while listing files in Google Drive: {e}")

if not apis_file_id:
    print(f"Error: Could not find APIs zip file with version {VERSION} in the specified folder.")
    sys.exit("Required APIs zip file not found.")

# 2. Download the found APIs zip file
print(f"Downloading APIs zip file with ID: {apis_file_id}...")
download_drive_file(drive_service, apis_file_id, ZIP_PATH, file_name=f'APIs_V{VERSION}.zip')

# 3. Extract specific items from the zip file to /content
print(f"Extracting specific items from {ZIP_PATH} to {CONTENT_DIR}...")
try:
    with zipfile.ZipFile(ZIP_PATH, 'r') as zip_ref:
        zip_contents = zip_ref.namelist()

        for member in zip_contents:
            extracted = False
            for item_prefix in ITEMS_TO_EXTRACT:
              if member == item_prefix or member.startswith(item_prefix):
                    zip_ref.extract(member, CONTENT_DIR)
                    extracted = True
                    break

except zipfile.BadZipFile:
    print(f"Error: The downloaded file at {ZIP_PATH} is not a valid zip file.")
    sys.exit("Invalid zip file downloaded.")
except Exception as e:
    print(f"An error occurred during extraction: {e}")
    sys.exit("Extraction failed.")


# 4. Clean up
if os.path.exists(ZIP_PATH):
    os.remove(ZIP_PATH)

# 5. Add APIs to path
if os.path.exists(APIS_DIR):
    sys.path.append(APIS_DIR)
else:
    print(f"Error: APIS directory not found at {APIS_DIR} after extraction. Cannot add to path.")

# 6. Quick verification
# Check for the presence of the extracted items
verification_paths = [APIS_DIR, DBS_DIR, SCRIPTS_DIR]
all_present = True
print("\nVerifying extracted items:")
for path in verification_paths:
    if os.path.exists(path):
        print(f"✅ {path} is present.")
    else:
        print(f"❌ {path} is MISSING!")
        all_present = False

if all_present:
    print(f"\n✅ Setup complete! Required items extracted to {CONTENT_DIR}.")
else:
    print("\n❌ Setup failed! Not all required items were extracted.")

# 7. Generate Schemas

print("\nGenerating FC Schemas")

# Change working directory to the source folder

# Iterate through the packages in the /content/APIs directory

    # Check if it's a directory (to avoid processing files)
        # Call the function to generate schema for the current package
print(f"✅ Successfully generated {len(os.listdir(FC_DIR))} FC Schemas to {FC_DIR}")
os.chdir(CONTENT_DIR)

Searching for APIs zip file with version 0.1.0 in folder: 1QpkAZxXhVFzIbm8qPGPRP1YqXEvJ4uD4...


Found matching file: APIs_V0.1.0.zip (ID: 1hLV2slrHhH0RquKU-8oWRJRs_nHh5CT_)


Download progress: 100%
Extracting specific items from /content/APIs_V0.1.0.zip to /content...



Verifying extracted items:
✅ /content/APIs is present.
✅ /content/DBs is present.
✅ /content/Scripts is present.

✅ Setup complete! Required items extracted to /content.



Generating FC Schemas
✅ instagram Schema generation complete: /content/Schemas/instagram.json


Processing mutation instagram.mutations.m01...
✅ instagram.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/instagram.json

✅ tiktok Schema generation complete: /content/Schemas/tiktok.json


Processing mutation tiktok.mutations.m01...
✅ tiktok.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/tiktok.json



✅ airline Schema generation complete: /content/Schemas/airline.json


Processing mutation airline.mutations.m01...
✅ airline.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/airline.json

✅ gemini_cli Schema generation complete: /content/Schemas/gemini_cli.json


Processing mutation gemini_cli.mutations.m01...
✅ gemini_cli.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/gemini_cli.json



✅ whatsapp Schema generation complete: /content/Schemas/whatsapp.json


Processing mutation whatsapp.mutations.m01...
✅ whatsapp.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/whatsapp.json

✅ google_maps_live Schema generation complete: /content/Schemas/google_maps_live.json


Processing mutation google_maps_live.mutations.m01...
✅ google_maps_live.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/google_maps_live.json



✅ notes_and_lists Schema generation complete: /content/Schemas/notes_and_lists.json


Processing mutation notes_and_lists.mutations.m01...
✅ notes_and_lists.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/notes_and_lists.json



✅ gdrive Schema generation complete: /content/Schemas/gdrive.json


Processing mutation gdrive.mutations.m01...
✅ gdrive.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/gdrive.json



✅ google_chat Schema generation complete: /content/Schemas/google_chat.json


Processing mutation google_chat.mutations.m01...
✅ google_chat.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/google_chat.json

✅ linkedin Schema generation complete: /content/Schemas/linkedin.json


Processing mutation linkedin.mutations.m01...
✅ linkedin.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/linkedin.json

✅ google_docs Schema generation complete: /content/Schemas/google_docs.json


Processing mutation google_docs.mutations.m01...
✅ google_docs.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/google_docs.json

✅ google_search Schema generation complete: /content/Schemas/google_search.json


Processing mutation google_search.mutations.m01...
✅ google_search.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/google_search.json



✅ device_setting Schema generation complete: /content/Schemas/device_setting.json


Processing mutation device_setting.mutations.m01...
✅ device_setting.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/device_setting.json

✅ blender Schema generation complete: /content/Schemas/blender.json


Processing mutation blender.mutations.m01...
✅ blender.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/blender.json



✅ azure Schema generation complete: /content/Schemas/azure.json


Processing mutation azure.mutations.m01...
✅ azure.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/azure.json

✅ bigquery Schema generation complete: /content/Schemas/bigquery.json


Processing mutation bigquery.mutations.m01...
✅ bigquery.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/bigquery.json

✅ contacts Schema generation complete: /content/Schemas/contacts.json


Processing mutation contacts.mutations.m01...
✅ contacts.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/contacts.json

✅ google_home Schema generation complete: /content/Schemas/google_home.json


Processing mutation google_home.mutations.m01...


✅ google_home.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/google_home.json

✅ youtube_tool Schema generation complete: /content/Schemas/youtube_tool.json


Processing mutation youtube_tool.mutations.m01...
✅ youtube_tool.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/youtube_tool.json



✅ google_people Schema generation complete: /content/Schemas/google_people.json


Processing mutation google_people.mutations.m01...
✅ google_people.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/google_people.json



✅ gmail Schema generation complete: /content/Schemas/gmail.json


Processing mutation gmail.mutations.m01...
✅ gmail.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/gmail.json

✅ generic_reminders Schema generation complete: /content/Schemas/generic_reminders.json


Processing mutation generic_reminders.mutations.m01...
✅ generic_reminders.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/generic_reminders.json

✅ authentication Schema generation complete: /content/Schemas/authentication.json


Processing mutation authentication.mutations.m01...
✅ authentication.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/authentication.json



✅ github_actions Schema generation complete: /content/Schemas/github_actions.json


Processing mutation github_actions.mutations.m01...
✅ github_actions.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/github_actions.json



✅ clock Schema generation complete: /content/Schemas/clock.json


Processing mutation clock.mutations.m01...
✅ clock.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/clock.json



✅ figma Schema generation complete: /content/Schemas/figma.json


Processing mutation figma.mutations.m01...
✅ figma.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/figma.json

Error: Could not find a valid _function_map in /content/APIs/common_utils/__init__.py.
✅ phone Schema generation complete: /content/Schemas/phone.json


Processing mutation phone.mutations.m01...
✅ phone.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/phone.json

✅ call_llm Schema generation complete: /content/Schemas/call_llm.json


Processing mutation call_llm.mutations.m01...
✅ call_llm.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/call_llm.json

✅ sdm Schema generation complete: /content/Schemas/sdm.json


Processing mutation sdm.mutations.m01...
✅ sdm.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/sdm.json

✅ google_slides Schema generation complete: /content/Schemas/google_slides.json


Processing mutation google

✅ spotify Schema generation complete: /content/Schemas/spotify.json


Processing mutation spotify.mutations.m01...
✅ spotify.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/spotify.json

✅ notifications Schema generation complete: /content/Schemas/notifications.json


Processing mutation notifications.mutations.m01...
✅ notifications.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/notifications.json

✅ canva Schema generation complete: /content/Schemas/canva.json


Processing mutation canva.mutations.m01...


✅ canva.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/canva.json



✅ slack Schema generation complete: /content/Schemas/slack.json


Processing mutation slack.mutations.m01...
✅ slack.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/slack.json



✅ reddit Schema generation complete: /content/Schemas/reddit.json


Processing mutation reddit.mutations.m01...


✅ reddit.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/reddit.json



✅ confluence Schema generation complete: /content/Schemas/confluence.json


Processing mutation confluence.mutations.m01...
✅ confluence.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/confluence.json



✅ shopify Schema generation complete: /content/Schemas/shopify.json


Processing mutation shopify.mutations.m01...
✅ shopify.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/shopify.json



✅ retail Schema generation complete: /content/Schemas/retail.json


Processing mutation retail.mutations.m01...
✅ retail.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/retail.json


Processing mutation retail.mutations.smaller_toolset...
✅ retail.mutations.smaller_toolset Schema generation complete: /content/MutationSchemas/smaller_toolset/retail.json



✅ zendesk Schema generation complete: /content/Schemas/zendesk.json


Processing mutation zendesk.mutations.m01...
✅ zendesk.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/zendesk.json



✅ workday Schema generation complete: /content/Schemas/workday.json


Processing mutation workday.mutations.m01...


✅ workday.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/workday.json

✅ service_template Schema generation complete: /content/Schemas/service_template.json


Processing mutation service_template.mutations.m01...
✅ service_template.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/service_template.json

✅ code_execution Schema generation complete: /content/Schemas/code_execution.json


Processing mutation code_execution.mutations.m01...
✅ code_execution.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/code_execution.json

✅ google_meet Schema generation complete: /content/Schemas/google_meet.json


Processing mutation google_meet.mutations.m01...
✅ google_meet.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/google_meet.json



✅ stripe Schema generation complete: /content/Schemas/stripe.json


Processing mutation stripe.mutations.m01...
✅ stripe.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/stripe.json

✅ mysql Schema generation complete: /content/Schemas/mysql.json


Processing mutation mysql.mutations.m01...
✅ mysql.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/mysql.json



✅ copilot Schema generation complete: /content/Schemas/copilot.json


Processing mutation copilot.mutations.m01...
✅ copilot.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/copilot.json



✅ cursor Schema generation complete: /content/Schemas/cursor.json


Processing mutation cursor.mutations.m01...
✅ cursor.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/cursor.json



✅ google_sheets Schema generation complete: /content/Schemas/google_sheets.json


Processing mutation google_sheets.mutations.m01...
✅ google_sheets.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/google_sheets.json

✅ google_maps Schema generation complete: /content/Schemas/google_maps.json


Processing mutation google_maps.mutations.m01...
✅ google_maps.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/google_maps.json



✅ device_actions Schema generation complete: /content/Schemas/device_actions.json


Processing mutation device_actions.mutations.m01...
✅ device_actions.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/device_actions.json

✅ sapconcur Schema generation complete: /content/Schemas/sapconcur.json


Processing mutation sapconcur.mutations.m01...
✅ sapconcur.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/sapconcur.json



✅ google_calendar Schema generation complete: /content/Schemas/google_calendar.json


Processing mutation google_calendar.mutations.m01...
✅ google_calendar.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/google_calendar.json



✅ puppeteer Schema generation complete: /content/Schemas/puppeteer.json


Processing mutation puppeteer.mutations.m01...
Error: Could not find a valid _function_map in /content/APIs/puppeteer/mutations/m01/__init__.py.


✅ github Schema generation complete: /content/Schemas/github.json


Processing mutation github.mutations.m01...
✅ github.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/github.json

✅ messages Schema generation complete: /content/Schemas/messages.json


Processing mutation messages.mutations.m01...
✅ messages.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/messages.json

✅ tool_explorer Schema generation complete: /content/Schemas/tool_explorer.json


Processing mutation tool_explorer.mutations.m01...
✅ tool_explorer.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/tool_explorer.json



✅ mongodb Schema generation complete: /content/Schemas/mongodb.json


Processing mutation mongodb.mutations.m01...
✅ mongodb.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/mongodb.json

✅ media_control Schema generation complete: /content/Schemas/media_control.json


Processing mutation media_control.mutations.m01...
✅ media_control.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/media_control.json



✅ salesforce Schema generation complete: /content/Schemas/salesforce.json


Processing mutation salesforce.mutations.m01...
✅ salesforce.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/salesforce.json



✅ jira Schema generation complete: /content/Schemas/jira.json


Processing mutation jira.mutations.m01...
✅ jira.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/jira.json

✅ terminal Schema generation complete: /content/Schemas/terminal.json


Processing mutation terminal.mutations.m01...
✅ terminal.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/terminal.json



✅ hubspot Schema generation complete: /content/Schemas/hubspot.json


Processing mutation hubspot.mutations.m01...
✅ hubspot.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/hubspot.json

✅ home_assistant Schema generation complete: /content/Schemas/home_assistant.json


Processing mutation home_assistant.mutations.m01...
✅ home_assistant.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/home_assistant.json



✅ supabase Schema generation complete: /content/Schemas/supabase.json


Processing mutation supabase.mutations.m01...
✅ supabase.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/supabase.json

✅ generic_media Schema generation complete: /content/Schemas/generic_media.json


Processing mutation generic_media.mutations.m01...
✅ generic_media.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/generic_media.json



✅ youtube Schema generation complete: /content/Schemas/youtube.json


Processing mutation youtube.mutations.m01...
✅ youtube.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/youtube.json



✅ google_cloud_storage Schema generation complete: /content/Schemas/google_cloud_storage.json


Processing mutation google_cloud_storage.mutations.m01...
✅ google_cloud_storage.mutations.m01 Schema generation complete: /content/MutationSchemas/m01/google_cloud_storage.json

✅ Successfully generated 67 FC Schemas to /content/Schemas


## Install Dependencies and Clone Repositories

In [ ]:
!pip install -r /content/APIs/requirements.txt

## Import APIs and initiate DBs

In [ ]:
# Using default DB for the Tau Benchmark
import retail

retail.SimulationEngine.db.load_state("/content/DBs/RetailDefaultDB.json")


# Initial Assertion

1. A user exists with email `fatima.wilson5721@example.com`.
2. An order with ID #W5272531 exists.
3. The status of order #W5272531 is 'delivered'.
4. Order #W5272531 contains items with the IDs '7228247242', '2698416822', '8098621301', and '3320557165'.
5. The user associated with order #W5272531 has a payment method with ID 'credit_card_6824399'.

In [ ]:
import retail
from Scripts.assertions_utils import *

# Define constants for the initial assertion
ORDER_ID = "#W5272531"
EXPECTED_STATUS = "delivered"
EXPECTED_ITEM_IDS = {'7228247242', '2698416822', '8098621301', '3320557165'}
EXPECTED_PAYMENT_METHOD_ID = "credit_card_6824399"
FIRST_NAME = "FATIMA"
LAST_NAME = "WILSON"
EMAIL = 'fatima.wilson5721@example.com'

# --- Data Gathering for Initial Assertions ---
order_details = None
user_details = None
api_call_error = None

try:
    # Fetch order details to verify its existence, status, and items
    order_details = retail.get_order_details(order_id=ORDER_ID)
    if order_details:
        # Fetch user details to verify payment methods
        user_id = order_details.get("user_id")
        if user_id:
            user_details = retail.get_user_details(user_id=user_id)
except Exception as e:
    api_call_error = str(e)


# --- Assertion 1: A user exists with email fatima.wilson5721@example.com. ---
assertion_condition_1 = False
try:
    user_id = retail.find_user_id_by_email(EMAIL)
except:
    pass

try:
    if user_id:
        assertion_condition_1 = True
except:
    pass

assert assertion_condition_1, "The user Fatima Wilson is not found in the database."

# --- Assertion 2: An order with ID #W5272531 exists. ---
assertion_message_2 = f"Assertion 2 Failed: Order with ID '{ORDER_ID}' was not found."
if api_call_error:
    assertion_message_2 += f" API call failed with error: {api_call_error}"
assertion_condition_2 = order_details is not None
assert assertion_condition_2, assertion_message_2

# --- Assertion 3: The status of order #W5272531 is 'delivered'. ---
actual_status = order_details.get("status", "Status not found")
assertion_message_3 = f"Assertion 3 Failed: Expected status for order '{ORDER_ID}' to be '{EXPECTED_STATUS}', but found '{actual_status}'."
assertion_condition_3 = compare_strings(actual_status,EXPECTED_STATUS)
assert assertion_condition_3, assertion_message_3

# --- Assertion 4: Order #W5272531 contains items with the IDs '7228247242', '2698416822', '8098621301', and '3320557165'. ---
items_in_order = order_details.get("items", [])
actual_item_ids = {item.get("item_id") for item in items_in_order if item.get("item_id")}
missing_items = EXPECTED_ITEM_IDS - actual_item_ids
assertion_message_4 = f"Assertion 4 Failed: Order '{ORDER_ID}' does not contain all expected items. Missing: {missing_items}. Found: {actual_item_ids}."
assertion_condition_4 = EXPECTED_ITEM_IDS.issubset(actual_item_ids)
assert assertion_condition_4, assertion_message_4

# --- Assertion 5: The user associated with order #W5272531 has a payment method with ID 'credit_card_6824399'. ---
user_payment_methods = {}
if user_details:
    user_payment_methods = user_details.get("payment_methods", {})
assertion_message_5 = f"Assertion 5 Failed: Payment method '{EXPECTED_PAYMENT_METHOD_ID}' not found for user '{order_details.get('user_id', 'N/A')}'. Available methods: {list(user_payment_methods.keys())}."
assertion_condition_5 = EXPECTED_PAYMENT_METHOD_ID in user_payment_methods
assert assertion_condition_5, assertion_message_5

# Action

**Simulated User**: I want to return everything I just bought except for the coffee machine.

Action Agent: Could you please provide me your email?

Simulated User: My email is fatima.wilson5721@example.com


In [ ]:
import retail

retail.find_user_id_by_email(email="fatima.wilson5721@example.com")

'fatima_wilson_7472'

In [ ]:
retail.get_user_details(user_id='fatima_wilson_7472')

{'name': {'first_name': 'Fatima', 'last_name': 'Wilson'},
 'address': {'address1': '167 Willow Lane',
  'address2': 'Suite 624',
  'city': 'San Diego',
  'country': 'USA',
  'state': 'CA',
  'zip': '92183'},
 'email': 'fatima.wilson5721@example.com',
 'payment_methods': {'credit_card_6824399': {'id': 'credit_card_6824399',
   'source': 'credit_card',
   'brand': 'visa',
   'last_four': '8991'}},
 'orders': ['#W5272531']}

In [ ]:
print(retail.get_order_details(order_id='#W5272531'))

{'order_id': '#W5272531', 'user_id': 'fatima_wilson_7472', 'address': {'address1': '167 Willow Lane', 'address2': 'Suite 624', 'city': 'San Diego', 'country': 'USA', 'state': 'CA', 'zip': '92183'}, 'items': [{'name': 'Electric Toothbrush', 'product_id': '7352963235', 'item_id': '3320557165', 'price': 188.67, 'options': {'color': 'blue', 'speed settings': 'high', 'battery type': 'AA batteries'}}, {'name': 'Electric Kettle', 'product_id': '1075968781', 'item_id': '2698416822', 'price': 149.45, 'options': {'capacity': '1.5L', 'material': 'plastic', 'color': 'white'}}, {'name': 'Electric Toothbrush', 'product_id': '7352963235', 'item_id': '8098621301', 'price': 192.15, 'options': {'color': 'black', 'speed settings': 'high', 'battery type': 'rechargeable'}}, {'name': 'Hiking Boots', 'product_id': '7363354090', 'item_id': '7228247242', 'price': 251.38, 'options': {'size': '10', 'material': 'leather', 'waterproof': 'yes'}}, {'name': 'Espresso Machine', 'product_id': '4354588079', 'item_id': '

**Action Agent**: I see your order includes an **Espresso Machine** (which you want to keep) and 4 other items you want to return:

* Electric Toothbrush (Blue) – \$188.67
* Electric Kettle (White) – \$149.45
* Electric Toothbrush (Black) – \$192.15
* Hiking Boots (Size 10) – \$251.38

Let me calculate your total refund.


In [ ]:
print(retail.calculate(expression='188.67 + 149.45 + 192.15 + 251.38'))

781.65


**Action Agent**: The total refund will be \$781.65, and it will be credited to your credit card (with last 4 digits as 8991). Do you want me to proceed?

**Simulated User**: Yes, please proceed.


In [ ]:
print(retail.return_delivered_order_items(
    order_id='#W5272531',
    item_ids=['3320557165', '2698416822', '8098621301', '7228247242'],  # All except coffee machine
    payment_method_id='credit_card_6824399'
))

{'order_id': '#W5272531', 'user_id': 'fatima_wilson_7472', 'address': {'address1': '167 Willow Lane', 'address2': 'Suite 624', 'city': 'San Diego', 'country': 'USA', 'state': 'CA', 'zip': '92183'}, 'items': [{'name': 'Electric Toothbrush', 'product_id': '7352963235', 'item_id': '3320557165', 'price': 188.67, 'options': {'color': 'blue', 'speed settings': 'high', 'battery type': 'AA batteries'}}, {'name': 'Electric Kettle', 'product_id': '1075968781', 'item_id': '2698416822', 'price': 149.45, 'options': {'capacity': '1.5L', 'material': 'plastic', 'color': 'white'}}, {'name': 'Electric Toothbrush', 'product_id': '7352963235', 'item_id': '8098621301', 'price': 192.15, 'options': {'color': 'black', 'speed settings': 'high', 'battery type': 'rechargeable'}}, {'name': 'Hiking Boots', 'product_id': '7363354090', 'item_id': '7228247242', 'price': 251.38, 'options': {'size': '10', 'material': 'leather', 'waterproof': 'yes'}}, {'name': 'Espresso Machine', 'product_id': '4354588079', 'item_id': '

**Action Agent**: Your return has been processed successfully. You’ll receive an email with instructions to send back the items. Your refund of **\$781.65** will be credited to your **credit card**. Is there anything else I can assist you with?


**Simulated User**: No, that’s all. Thank you!

**Action Agent**: You’re welcome! Have a great day.


# Final Assertion

1. The order `#W5272531` is successfully updated, and its status now shows `return requested`.

In [ ]:
import retail
import re
from Scripts.assertions_utils import *

# =========================
# Constants / Expectations
# =========================
ORDER_ID = "#W5272531"
EXPECTED_ORDER_STATUS = "return requested"            # order-level status
EXPECTED_PAYMENT_METHOD = "credit_card_6824399"       # id or token of payment method
EXPECTED_ITEM_COUNT = 4                               # all items expected to be returned
# Which item-level statuses count as "returned" for the purpose of refund?
RETURNED_STATUSES = {"returned", "return requested", "refunded"}

# ================
# Helper Functions
# ================
def get_safe(dct, *path, default=None):
    """Safe nested getter: get_safe(d, 'a','b','c', default=None)."""
    cur = dct
    for key in path:
        if not isinstance(cur, dict) or key not in cur:
            return default
        cur = cur[key]
    return cur

def normalize_str(s):
    return s.strip().lower() if isinstance(s, str) else s

# Fetch current order details
order_details = retail.get_order_details(order_id=ORDER_ID)

# ------------------------------------------------------------
# Assertion 1: Order status shows expected "return requested".
# ------------------------------------------------------------
actual_status = normalize_str(order_details.get("status"))
assertion_message_1 = (
    f"Assertion Failed: Expected order status '{EXPECTED_ORDER_STATUS}', "
    f"got '{actual_status}'."
)
assert compare_strings(actual_status, EXPECTED_ORDER_STATUS), assertion_message_1

# -------------------------------------------------------------------
# Assertion 2: All items (expected 4) are marked as returned/requested.
# -------------------------------------------------------------------
items = order_details.get("items") or []
returned_like = [
    it for it in items
    if normalize_str(it.get("status")) in RETURNED_STATUSES
]
assertion_message_2 = (
    f"Assertion Failed: Expected {EXPECTED_ITEM_COUNT} items to be in a returned "
    f"state ({', '.join(sorted(RETURNED_STATUSES))}), "
    f"but found {len(returned_like)} out of {len(items)}."
)
assert len(returned_like) == EXPECTED_ITEM_COUNT, assertion_message_2

# --------------------------------------------------------------------------------------
# Assertion 3: Refund(s) were issued to the correct payment method AND amount matches.
# --------------------------------------------------------------------------------------
# Assumptions about schema (robust to variation):
# - order_details['refunds'] -> list of refund objects with keys: 'amount', 'payment_method'
# - Each returned item may have 'refund_amount' (prefer) or fallback to item 'total'/'price'.
refunds = order_details.get("refunds") or []
payment_method_ok = all(
    normalize_str(r.get("payment_method")) == normalize_str(EXPECTED_PAYMENT_METHOD)
    for r in refunds
) if refunds else False

assertion_message_3a = (
    "Assertion Failed: Refunds were not issued to the expected payment method "
    f"'{EXPECTED_PAYMENT_METHOD}'. Refunds observed: "
    f"{[r.get('payment_method') for r in refunds] if refunds else 'NONE'}"
)
assert payment_method_ok, assertion_message_3a

# Compute expected refund from item-level data
def item_expected_refund(item):
    # prefer explicit refund_amount if present
    if isinstance(item.get("refund_amount"), (int, float)):
        return float(item["refund_amount"])
    # fallback: use 'total' or 'price' if available
    for key in ("total", "price", "unit_price"):
        if isinstance(item.get(key), (int, float)):
            # If quantity exists and total not present, multiply unit price * qty
            if key == "unit_price" and isinstance(item.get("quantity"), (int, float)):
                return float(item["unit_price"]) * float(item["quantity"])
            return float(item[key])
    return 0.0

expected_refund_amount = sum(item_expected_refund(it) for it in returned_like)

# Actual refund amount observed (sum of all refunds)
actual_refund_amount = sum(
    float(r.get("amount", 0.0)) for r in refunds
)

# Allow tiny float drift
EPS = 1e-6
amount_ok = abs(actual_refund_amount - expected_refund_amount) <= EPS

assertion_message_3b = (
    "Assertion Failed: Refund amount mismatch. "
    f"Expected total refund based on returned items = {expected_refund_amount:.2f}, "
    f"but observed refund total = {actual_refund_amount:.2f}."
)
assert amount_ok, assertion_message_3b

# ---------------------------------------------------------
# Assertion 4: Customer email is not a placeholder domain.
# ---------------------------------------------------------
email = get_safe(order_details, "customer", "email") or ""
email_str = str(email).strip()

# Basic email pattern and explicit placeholder domain block
email_pattern = r"^[^@\s]+@[^@\s]+\.[^@\s]+$"
uses_placeholder = email_str.lower().endswith("@example.com")

assertion_message_4a = (
    f"Assertion Failed: Invalid or missing customer email: '{email_str}'."
)
assert re.match(email_pattern, email_str or "") is not None, assertion_message_4a

assertion_message_4b = (
    "Assertion Failed: Customer email uses placeholder domain 'example.com' "
    f"(found '{email_str}')."
)
assert not uses_placeholder, assertion_message_4b

# ---------------------------------------------------------
# Optional sanity: Verify payment method on the order itself
# ---------------------------------------------------------
order_payment_method = normalize_str(
    get_safe(order_details, "payment", "method_id", default=order_details.get("payment_method"))
)
if order_payment_method is not None:
    # Only assert if present to avoid false negatives on missing data
    assertion_message_5 = (
        "Assertion Failed: Order's recorded payment method does not match the expected "
        f"one. Expected '{EXPECTED_PAYMENT_METHOD}', got '{order_payment_method}'."
    )
    assert compare_strings(order_payment_method, EXPECTED_PAYMENT_METHOD), assertion_message_5
